In [16]:
import pandas as pd
df_sales = pd.read_csv('csv/sales.csv')
print(df_sales.shape)
print(df_sales.head())
df_customers = pd.read_csv('csv/customers.csv')
print(df_customers.shape)
print(df_customers.head())
df = pd.merge(df_sales, df_customers, on='customer_id', how='left')
print(df.shape)
print(df.dtypes)
print(df.head())

(15234, 11)
   sale_id        invoice_id   sale_date  customer_id  product_id  quantity  \
0        1  INV-202301-00001  2023-01-10       2588.0        1274         2   
1        2  INV-202301-00002  2023-01-16       4338.0        1207         2   
2        3  INV-202301-00003  2023-01-05       4673.0        1264         1   
3        4  INV-202301-00004  2023-01-02       4677.0        1341         3   
4        5  INV-202301-00005  2023-01-13       2390.0        1284         1   

   unit_price  total_price channel store_location payment_method  
0      234.79       469.58    pood        Tallinn          kaart  
1      241.13       482.26    pood          Pärnu      järelmaks  
2      258.46       221.19    pood          Pärnu      järelmaks  
3       45.21       135.63    pood          Tartu       sularaha  
4       99.57        99.57    pood          Tartu          kaart  
(3150, 9)
   customer_id first_name last_name                   email           phone  \
0         2001        

In [ ]:
df.isna().sum()

In [5]:
df.isna().sum()

sale_id                 0
invoice_id              0
sale_date               0
customer_id          1487
product_id              0
quantity                0
unit_price              0
total_price             0
channel                 0
store_location       5204
payment_method          0
first_name           1487
last_name            1487
email                2905
phone                1487
city                 1487
registration_date    1487
loyalty_tier         7033
birth_year           1487
dtype: int64

In [6]:
df.duplicated().sum()

np.int64(4086)

In [7]:
df[df.duplicated()].head()

,sale_id,invoice_id,sale_date,customer_id,product_id,quantity,unit_price,total_price,channel,store_location,payment_method,first_name,last_name,email,phone,city,registration_date,loyalty_tier,birth_year
10118,6957,INV-202407-00266,2024-07-05,4909.0,1084,1,156.13,156.13,pood,Tallinn,järelmaks,Mihkel,Talvik,mihkel.talvik2@hot.ee,+372 5128 9327,Tartu,2021-05-12,silver,1996.0
10120,506,INV-202302-00245,2023-02-16,4307.0,1315,1,95.50,95.50,pood,Pärnu,kaart,Merike,Rebane,merike.rebane@outlook.com,+372 5126 9930,Kuressaare,2022-07-01,NaN,1998.0
10121,3791,INV-202311-00286,2023-11-29,4701.0,1343,3,63.87,191.61,pood,Tallinn,järelmaks,Enn,Rebane,NaN,+372 5952 5043,Tartu,2022-12-16,NaN,1967.0
10122,6982,INV-202407-00291,2024-07-30,3868.0,1307,3,127.38,382.14,pood,Tallinn,sularaha,Anu,Paju,anu.paju@yahoo.com,+372 8407 9765,Tallinn,2020-09-30,NaN,1978.0
10125,2641,INV-202308-00218,2023-08-15,3964.0,1197,2,152.85,305.70,pood,Tartu,kaart,Kersti,Männik,kersti.mannik@yahoo.com,+372 5850 6802,Tartu,2022-04-10,bronze,1980.0


In [8]:
df = df.drop_duplicates()

In [9]:
df.shape

(11148, 19)

In [10]:
df['sale_date'] = pd.to_datetime(df['sale_date'])

ValueError: time data "25/01/2023" doesn't match format "%Y-%m-%d". You might want to try:
    - passing `format` if your strings have a consistent format;
    - passing `format='ISO8601'` if your strings are all ISO8601 but not necessarily in exactly the same format;
    - passing `format='mixed'`, and the format will be inferred for each element individually. You might want to use `dayfirst` alongside this.

In [11]:
df['sale_date'] = pd.to_datetime(
    df['sale_date'],
    format='mixed',
    dayfirst=True
)

In [12]:
df['sale_date'].dtype

dtype('<M8[us]')

In [13]:
df['customer_id'].isna().sum()

np.int64(1236)

In [14]:
df = df.dropna(subset=['customer_id'])

In [15]:
df.shape

(9912, 19)

In [17]:
df.duplicated(subset=['invoice_id']).sum()

np.int64(5116)

In [18]:
df = df.drop_duplicates(subset=['invoice_id'], keep='first')

In [19]:
df.shape

(10118, 19)

In [20]:
df.isnull().sum()

sale_id                 0
invoice_id              0
sale_date               0
customer_id           988
product_id              0
quantity                0
unit_price              0
total_price             0
channel                 0
store_location       3462
payment_method          0
first_name            988
last_name             988
email                1944
phone                 988
city                  988
registration_date     988
loyalty_tier         4660
birth_year            988
dtype: int64

In [21]:
df = df.dropna(subset=['customer_id', 'sale_date', 'total_price'])

In [22]:
df['sale_date'] = pd.to_datetime(
    df['sale_date'],
    format='mixed',
    dayfirst=True
)

In [23]:
(df['total_price'] < 0).sum()

np.int64(180)

In [24]:
df = df[df['total_price'] > 0]

In [25]:
df.shape

(8950, 19)

In [26]:
df['customer_id'].nunique()

2540

In [27]:
df['sale_date'].min(), df['sale_date'].max()

(Timestamp('2023-01-01 00:00:00'), Timestamp('2026-12-03 00:00:00'))

## Roll B – Andmete puhastamine

Puhastamise käigus:
- eemaldasin 5116 duplikaati `invoice_id` järgi;
- eemaldasin 988 rida, kus puudus `customer_id`;
- teisendasin `sale_date` kuupäevaformaati;
- eemaldasin 180 negatiivse `total_price` väärtusega rida.

Pärast puhastamist:
- ridu: 8950
- veerge: 19
- unikaalseid kliente: 2540
- kuupäevavahemik: 01.01.2023–03.12.2026

Puhastatud DataFrame on valmis RFM-analüüsiks.